### DISCUSSION TITLE
How to build fwupdtool for macOS

### DISCUSSION THREAD

**Discussion by GoodWayDev** [fwupd/fwupd#6041] [Q&A]
Hi, I want to compile the fwupd executable (similar to an .exe file on Windows) on my MacBook M1. How can I do that?

After trying to execute ./contrib/ci/build_macos.sh, it generates a build-macos folder in the root directory, but I can't find any executable file in that directory.

**Comment by hughsie** [comment]
In short: I'm not really sure :) I don't have any macOS hardware -- and the only way I test building for macOS is in CI. I do think you need to have a developer account to build the binaries for installation but I've never done it myself. Sorry that's not more helpful.

**Comment by GoodWayDev** [comment]
Thanks for your reply, I found the fwupdtool executable in the build-macos/src folder, which is what I was looking for.
But I get a failure message when executing fwupdtool get-devices, do you have any experience?
<img width="682" alt="Failed to create '/run/lock': Read-only file system" src="https://github.com/fwupd/fwupd/assets/72374862/f7184a7b-1598-4ab4-8d5e-00c203671183">

**Comment by hughsie** [comment]
I'm afraid I don't know; what is the lock directory for macos? Does `/run/lock` even exist? Does `sudo FWUPD_LOCKDIR=/var/run ./fwupdtool get-devices` work?

**Reply by GoodWayDev** [reply]
`sudo FWUPD_LOCKDIR=/var/run ./fwupdtool get-devices` looks fine, it starts the process, but there is a new error, can it be ignored?
[GLib failed.pdf](https://github.com/fwupd/fwupd/files/12213646/GLib.failed.pdf)

**Reply by hughsie** [reply]
Yes, you can ignore that for now. Out of interest, does /run/lock even exist?

**Reply by GoodWayDev** [reply]
I don't know what `/run/lock` is?  How can I check if it exists?

**Reply by hughsie** [reply]
e.g. what does `ls /run/lock` say?

**Reply by GoodWayDev** [reply]
No, it doesn't exist, showing `ls: /run/lock: No such file or directory`

**Reply by hughsie** [reply]
Okay, https://github.com/fwupd/fwupd/pull/6049 will help then -- thanks.

**Comment by rokorec31** [comment]
I checked out the latest master branch (commit 8251152), but still can't get the device list. Can you please verify the logs I got. Is there anything wrong?
[20230801_get-devices_log.pdf](https://github.com/fwupd/fwupd/files/12225327/20230801_get-devices_log.pdf)

I have verified v1.9.3 on windows and it recognizes my device Gingerbread.

<img width="1137" alt="IOUSB" src="https://github.com/fwupd/fwupd/assets/10734560/8bd2558f-cfd7-4d9c-a58a-64187cbeb04a">

**Reply by hughsie** [reply]
What's the output of `sudo ./fwupdtool get-devices -vv` please?

**Reply by rokorec31** [reply]
Here's the output
[20230810_get-devices_log.pdf](https://github.com/fwupd/fwupd/files/12309874/20230810_get-devices_log.pdf)

**Comment by hughsie** [comment]
I've spent a bit of time trying to reproduce this myself (which included ~1d of installing Xcode on a very very old macbook :)

There is one fix needed in libxmlb: https://github.com/hughsie/libxmlb/pull/141/commits/1208904fa1148d6972051e04d2b6d2cf04702ee6

And another set of fixes needed in https://github.com/fwupd/fwupd/pull/6068

if you `git checkout master` in the libxmlb subproject and switch to the `wip/hughsie/macos-kernel_cmdline` fwupd branch it should work. You might want to wait for the next release as that'll just work.

<img width="615" alt="Screenshot 2023-08-10 at 10 16 10" src="https://github.com/fwupd/fwupd/assets/151380/bf389356-e3b5-4845-a245-dfa82937e62e">
